lib/experimental/html/class-wp-html-tag-processor.php

```diff
@@ -1397,6 +1397,24 @@ public function get_attribute( $name ) {
 		}
 
 		$comparable = strtolower( $name );
+
+		if ( 'class' === $name ) {
+			$this->class_name_updates_to_attributes_updates();
+		}
```

Comment:
For `$this->attribute_updates` (below) it makes sense (and is easy enough) to only evaluate the updates for the attribute we're interested in (if any) and to ignore updates for all other attributes. For class name updates OTOH, we would need to replicate most of the logic from `class_name_updates_to_attributes_updates` here, so we might as well just call that function and have classname updates carried over to attribute updates.

cc/ @adamziel since the PHPDoc for `class_name_updates_to_attributes_updates` says 

```
	 * This method is only meant to run right before the attribute updates are applied.
	 * The behavior in all other cases is undefined.
```

Although it looks okay to me to call it here?

Reply:
I think we need to call `$this->apply_attribute_updates()` here as well to ensure they get applied before any successive `class` updates occur, lest we lose enqueued class name updates.

Reply:
Ah yeah, you're right! While `class_name_updates_to_attributes_updates` does take an existing `class` attribute into account, it only looks at `$this->attributes['class']` to do so -- but not into `$lexical_updates`: https://github.com/WordPress/gutenberg/blob/777eaf2a34d11d3cce8456db85151141cc95a071/lib/experimental/html/class-wp-html-tag-processor.php#L1134-L1136

So we do need to call `$this->apply_attribute_updates()` as you say (unless we wanna change `class_name_updates_to_attributes_updates` to take `$lexical_updates['class']` into account -- but I guess we'd rather not).

I'll write a test case and will add a `$this->apply_attribute_updates()` call 👍

Reply:
Note that this will change existing semantics of `add_class` and `set_attribute` -- see: https://github.com/WordPress/gutenberg/blob/777eaf2a34d11d3cce8456db85151141cc95a071/phpunit/html/wp-html-tag-processor-test.php#L796-L828

Adding the `apply_attribute_updates` call thus breaks that test (plus a few other ones, going to look into those now).

Reply:
<details>
<summary>
Considering something like this
</summary>

```diff
diff --git a/lib/experimental/html/class-wp-html-tag-processor.php b/lib/experimental/html/class-wp-html-tag-processor.php
index e9c42bf419..9edd6a4105 100644
--- a/lib/experimental/html/class-wp-html-tag-processor.php
+++ b/lib/experimental/html/class-wp-html-tag-processor.php
@@ -1126,14 +1126,24 @@ class WP_HTML_Tag_Processor {
         * @see $lexical_updates
         */
        private function class_name_updates_to_attributes_updates() {
-               if ( count( $this->classname_updates ) === 0 || isset( $this->lexical_updates['class'] ) ) {
+               if ( count( $this->classname_updates ) === 0 ) {
                        $this->classname_updates = array();
                        return;
                }
 
-               $existing_class = isset( $this->attributes['class'] )
-                       ? substr( $this->html, $this->attributes['class']->value_starts_at, $this->attributes['class']->value_length )
-                       : '';
+               if ( isset( $this->lexical_updates['class'] ) ) {
+                       $existing_class_attr  = trim( $this->lexical_updates['class']->text );
+                       $existing_class_value = substr( $existing_class_attr, strlen( 'class' ) + 2, -1 );
+                       $existing_class       = html_entity_decode( $existing_class_value );
+               } elseif ( isset( $this->attributes['class'] ) ) {
+                       $existing_class = substr(
+                               $this->html,
+                               $this->attributes['class']->value_starts_at,
+                               $this->attributes['class']->value_length
+                       );
+               } else {
+                       $existing_class = '';
+               }
 
                /**
                 * Updated "class" attribute value.
```
</details>

But that will also change the semantics (two test cases -- both of which are about the `add_class`/`set_attribute` interaction AFAICS).

Reply:
> Adding the `apply_attribute_updates` call thus breaks that test (plus a few other ones, going to look into those now).

So the other tests affected are (the also fairly recently added) [`test_get_attribute_reflects_added_class_names_before_they_are_updated`](https://github.com/WordPress/gutenberg/pull/46680/files#diff-ea212be64658e9e6c7c4a6ac779899518ba7c773f4b4aeb8e022889274f0b6d4R551-R557), [`test_get_attribute_reflects_added_class_names_before_they_are_updated_and_retains_classes_from_previous_add_class_calls`](https://github.com/WordPress/gutenberg/pull/46680/files#diff-ea212be64658e9e6c7c4a6ac779899518ba7c773f4b4aeb8e022889274f0b6d4R566-R574) , and [`test_get_attribute_reflects_removed_class_names_before_they_are_updated`](https://github.com/WordPress/gutenberg/pull/46680/files#diff-ea212be64658e9e6c7c4a6ac779899518ba7c773f4b4aeb8e022889274f0b6d4R615-R621).

All of them call `get_attribute( 'class' )` right after setting (or removing) a class. Turns out that the problem is that if we add `$this->apply_attributes_updates();`, this flushes updates (so `get_attribute` doesn't pick up the newly added class names from `$this->lexical_attributes`; but it also doesn't parse the newly added attributes into `$this->attributes`, so `get_attribute` can't find them there either 😬 

Getting a bit into the weeds of all that update management I guess. I'm still leaning towards [this other solution](https://github.com/WordPress/gutenberg/pull/46680/files#r1072477489), which I believe has a smaller overall footprint.

Reply:
I like how you managed to avoid applying the attributes updates here 👍

Reply:
> an obvious outcome of calling add_class before the transactional swap of set_attribute

It is confusing 🤔 I'd like `set_attribute` to cancel any `add_class`/`remove_class` calls that happened before. I'd be fine with an undefined behavior accompanied by warning, too, though.

Reply:
@adamziel in my latest commit I restored that defined behavior. in other words, I'm currently thinking the change in behavior wasn't necessary and so I've removed it.

Reply:
Thanks for clarifying @ockham - you were right, I didn't understand that this was broken before the PR. The intention was right, the implementation was wrong.